Спринт 12

PapaRimsky opened the pull request:
Не понимаю почему некорректно работает показ списка треков по запросу(из 11 спринта). Первый запрос по апи работает хорошо, а следующие нет. Приложение получает список треков, но они не отображаются. Эта проблема возникает из-за функции loadTracks, а именно из-за присвоения списку треков адаптера список из памяти. Вообще не понимаю как это связано. Нужна помощь(

PapaRimsky committed:
Спринт 12 ошибкой

PapaRimsky committed:
Спринт 12 ошибкой

PapaRimsky committed:
Новые цвета для switch

PapaRimsky committed:
Новые цвета для switch

PapaRimsky committed:
12 спринт

nadzyathedude commented on app/src/main/java/com/paparimsky/playlistmaker/App.kt:
```diff
@@ -0,0 +1,52 @@
+        private const val LIGHT_THEME = "light"
+        private const val DARK_THEME = "dark"
+        var darkTheme = false
+        var sharedPrefs: SharedPreferences? = null
```
предлагаю избавиться от типа nullable, можно сделать  `lateinit var sharedPrefs: SharedPreferences`

nadzyathedude commented on app/src/main/java/com/paparimsky/playlistmaker/SearchActivity.kt:
```diff
@@ -35,46 +36,51 @@ class SearchActivity : AppCompatActivity() {
         super.onCreate(savedInstanceState)
         binding = ActivitySearchBinding.inflate(layoutInflater)
         setContentView(binding?.root)
+        binding?.backToMainFromSearch?.setOnClickListener {
```
в активити view binding можно смело делать не nullable

nadzyathedude commented on app/src/main/java/com/paparimsky/playlistmaker/SearchActivity.kt:
```diff
@@ -84,6 +90,31 @@ class SearchActivity : AppCompatActivity() {
+            }
+        }
+
+        App.sharedPrefs?.registerOnSharedPreferenceChangeListener(listener)
```
Всегда важно помнить что у приложения есть свой жизенный цикл. Важно зарегистрировать слушатель в верной месте, а еще важно не забыть отключить его, иначе может произойти утечка памяти. Вызови метод registerOnSharedPreferenceChangeListener(listener) в   onResume() и не забудь вызвать unregisterOnSharedPreferenceChangeListener(this) в onPause()

nadzyathedude commented on app/src/main/java/com/paparimsky/playlistmaker/SearchActivity.kt:
```diff
@@ -84,6 +90,31 @@ class SearchActivity : AppCompatActivity() {
+                binding?.searchBefore?.visibility = View.GONE
+            }
+        }
+        binding?.clearHistory?.setOnClickListener{
+            App.sharedPrefs
+                ?.edit()
+                ?.remove(App.SEARCHED_TRACKS_KEY)
+                ?.apply()
+            binding?.searchBefore?.visibility = View.GONE
+        }
```
предлагаю выносить отдельные состояния экрана в отдельные функции

nadzyathedude commented on app/src/main/java/com/paparimsky/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,34 @@
+    fun getHistory(): ArrayList<Track> {
+        val type = object : TypeToken<ArrayList<Track>>() {}.type
+        val jsonString = sharedPrefs?.getString(App.SEARCHED_TRACKS_KEY, null)
+        return Gson().fromJson(jsonString, type)
```
на этой строчке приложение падает с ошибкой 
`java.lang.RuntimeException: Unable to start activity ComponentInfo{com.paparimsky.playlistmaker/com.paparimsky.playlistmaker.SearchActivity}: java.lang.NullPointerException: Gson().fromJson(jsonString, type) must not be null`

PapaRimsky committed:
1) По поводу lateinit мне ранее оставлял комментарий ревьюер, сказал, что лучше использовать null, так безопаснее, и есть вариант забыть проинициализировать переменную или обратиться и тогда оно вылезет в исключение. Правильно ли так размышлять с вашей точки зрения?
2) По поводу binding, так же вопрос, разве с ?. не безопаснее? И почему все же стоит использовать !! смело?
3) Ошибку NullPointer исправил, но все равно остался вопрос по поводу, как эта функция влияет на отображение списка треков по запросу. После первого запроса, список больше не отображается.

PapaRimsky committed:
1) По поводу lateinit мне ранее оставлял комментарий ревьюер, сказал, что лучше использовать null, так безопаснее, и есть вариант забыть проинициализировать переменную или обратиться и тогда оно вылезет в исключение. Правильно ли так размышлять с вашей точки зрения?
2) По поводу binding, так же вопрос, разве с ?. не безопаснее? И почему все же стоит использовать !! смело?
3) Ошибку NullPointer исправил, но все равно остался вопрос по поводу, как эта функция влияет на отображение списка треков по запросу. После первого запроса, список больше не отображается.

nadzyathedude commented:
обычно binding выносят в lateinit  и инициализируют в onCreate() или же  лениво инициализируют с помощью конструкции lazy. а вот оператор !! наоборот крайне не рекомендуется использовать в целом везде. если помнишь, используя !! мы говорим компилятору что ну уж точно не будет значения null у переменной, а может быть всякое.  с binding  в активити можем быть уверены что null не будет, но тут есть одно но. далее ты изучишь такую штуку, как фрагменты. вот там обязательно делать binding nullable, так как придется давать ему значение null в одном из методов жизненного цикла. 
всегда полезно обратиться к [официальной документации ](https://developer.android.com/topic/libraries/view-binding#activities
)

nadzyathedude commented:
по поводу твоего непонимания с lateinit. вообще, на сама деле вкусовщина каждого, кто-то считает что его лучше не трогать, кому-то этот подход наоборот ближе. да, ты прав что можно забыть про инициализацию и вылезет ошибка. предлагаю тогда тебе действовать подходом, который тебе ближе

nadzyathedude commented on app/src/main/java/com/paparimsky/playlistmaker/SearchActivity.kt:
```diff
@@ -169,4 +198,14 @@ class SearchActivity : AppCompatActivity() {
+
+    @SuppressLint("NotifyDataSetChanged")
+    private fun loadTracks() {
+        searchedAdapter.tracksSearched = searchHistory.getHistory()
```
судя по дебагу, эта строка кода не вызывается

nadzyathedude commented:
думаю стоило обратиться к наставнику еще до первой итерации с твоим багом. предлагаю сейчас написать наставнику, вместе попробовать решить проблему и сдать работу

PapaRimsky committed:
В нужном месте сделал у списка треков view.visible

PapaRimsky committed:
Изменил binding с nullable на lateinit
